internal/controllers/lvmcluster/status.go

```diff
@@ -0,0 +1,161 @@
+	}
+	if init {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonReconciliationInProgress
+		condition.Message = MessageReconciliationInProgress
+		return condition
+	}
+
+	if reconciliationError == nil {
+		condition.Status = corev1.ConditionTrue
+		condition.Reason = ReasonResourcesSynced
+		condition.Message = MessageResourcesSynced
+	} else {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonResourcesSyncFailed
+		condition.Message = fmt.Sprintf(MessageReasonResourcesSyncFailed, reconciliationError)
+	}
+
+	return condition
+}
+
+func computeVGReadinessCondition(init bool, vgNodeStatusList *lvmv1alpha1.LVMVolumeGroupNodeStatusList) lvmv1alpha1.Condition {
+	condition := lvmv1alpha1.Condition{
+		Type: lvmv1alpha1.VolumeGroupsReady,
+	}
+	if init {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonVGReadinessInProgress
+		condition.Message = MessageVGReadinessInProgress
+		return condition
+	}
+
+	var isReady, isDegraded, isFailed, isProgressing bool
+	for _, nodeItem := range vgNodeStatusList.Items {
+		for _, item := range nodeItem.Spec.LVMVGStatus {
+			if item.Status == lvmv1alpha1.VGStatusReady {
+				isReady = true
+			} else if item.Status == lvmv1alpha1.VGStatusDegraded {
+				isDegraded = true
+			} else if item.Status == lvmv1alpha1.VGStatusFailed {
+				isFailed = true
+			} else if item.Status == lvmv1alpha1.VGStatusProgressing {
+				isProgressing = true
+			}
+		}
+	}
+
+	if isFailed {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonVGsFailed
+		condition.Message = MessageVGsFailed
+	} else if isDegraded {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonVGsDegraded
+		condition.Message = MessageVGsDegraded
+	} else if isProgressing {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = ReasonVGReadinessInProgress
+		condition.Message = MessageVGReadinessInProgress
+	} else if isReady {
+		condition.Status = corev1.ConditionTrue
+		condition.Reason = ReasonVGsReady
+		condition.Message = MessageVGsReady
+	}
+
+	return condition
+}
+
+func computeDeviceClassStatuses(vgNodeStatusList *lvmv1alpha1.LVMVolumeGroupNodeStatusList) []lvmv1alpha1.DeviceClassStatus {
+	vgNodeMap := make(map[string][]lvmv1alpha1.NodeStatus)
+	for _, nodeItem := range vgNodeStatusList.Items {
+		for _, item := range nodeItem.Spec.LVMVGStatus {
+			vgNodeMap[item.Name] = append(vgNodeMap[item.Name],
+				lvmv1alpha1.NodeStatus{
+					Node:     nodeItem.Name,
+					VGStatus: *item.DeepCopy(),
+				},
+			)
+		}
+	}
+	var allVgStatuses []lvmv1alpha1.DeviceClassStatus
+	for key, val := range vgNodeMap {
+		allVgStatuses = append(allVgStatuses,
+			lvmv1alpha1.DeviceClassStatus{
+				Name:       key,
+				NodeStatus: val,
+			},
+		)
+	}
+
+	return allVgStatuses
+}
+
+func computeReadiness(conditions []lvmv1alpha1.Condition) (state lvmv1alpha1.LVMStateType, ready bool) {
+	readyConditionCount := 0
+	for _, c := range conditions {
```

Comment:
similarly to above but a bit more complex, You could just translate c.Reason to a State in a function and then this function would get significantly less complex